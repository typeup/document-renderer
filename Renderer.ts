import * as dom from "@typeup/dom"
export type renderFunction = (renderer: Renderer, node: dom.Node) => Promise<string>

export class Renderer {
	private constructor(private variables: { [name: string]: string } = {}) {
	}
	getVariable(name: string): string {
		return this.variables[name]
	}
	setVariable(name: string, value: string) {
		this.variables[name] = value
	}
	async render(node: dom.Node | dom.Node[]): Promise<string> {
		let result: string
		if (Array.isArray(node))
			result = (await Promise.all(node.map(n => this.render(n)))).reduce<string[]>((r, e) => Array.isArray(e) ? [...r, ...e] : [...r, e], []).join("")
		else {
			const render = renderers[node.class]
			result = render ? await render(this, node) : ""
		}
		return result
	}
	scope(): Renderer {
		return new Renderer({ ...this.variables })
	}
	static create(): Renderer {
		return new Renderer({})
	}
}
const renderers: { [className: string]: renderFunction } = {}
export function addRenderer(className: string, render: renderFunction) {
	renderers[className] = render
}
