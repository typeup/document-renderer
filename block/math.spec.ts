import * as dom from "@typeup/dom"
import { Renderer } from "../Renderer"
import "../index"

const renderer = Renderer.create()
describe("Block.Math", () => {
	it("render", async () => {
		const node = new dom.block.Math("a^2 + b^2 = c^2", [new dom.inline.Text("Pythagoras Theorem.")])
		expect(await renderer.render(node)).toEqual(
`<figure>
	<math xmlns="http://www.w3.org/1998/Math/MathML" alttext="a^2 + b^2 = c^2">
  <msup>
    <mi>a</mi>
    <mn>2</mn>
  </msup>
  <mo>+</mo>
  <msup>
    <mi>b</mi>
    <mn>2</mn>
  </msup>
  <mo>=</mo>
  <msup>
    <mi>c</mi>
    <mn>2</mn>
  </msup>
</math>
\t<figcaption>Pythagoras Theorem.</figcaption>
</figure>
`)
	})
	it("render two", async () => {
		const node = new dom.block.Math("a^2 + b^2 = c^2", [new dom.inline.Text("Pythagoras Theorem.")])
		expect(await renderer.render([node, node])).toEqual(
`<figure>
	<math xmlns="http://www.w3.org/1998/Math/MathML" alttext="a^2 + b^2 = c^2">
  <msup>
    <mi>a</mi>
    <mn>2</mn>
  </msup>
  <mo>+</mo>
  <msup>
    <mi>b</mi>
    <mn>2</mn>
  </msup>
  <mo>=</mo>
  <msup>
    <mi>c</mi>
    <mn>2</mn>
  </msup>
</math>
\t<figcaption>Pythagoras Theorem.</figcaption>
</figure>
`.repeat(2))
	})
})
