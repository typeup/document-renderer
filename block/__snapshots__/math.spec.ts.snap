// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Block.Math render 1`] = `"<figure>$$a^2 + b^2 = c^2$$<figcaption>Pythagoras Theorem.</figcaption></figure>"`;

exports[`Block.Math render two 1`] = `"<figure>$$a^2 + b^2 = c^2$$<figcaption>Pythagoras Theorem.</figcaption></figure><figure>$$a^2 + b^2 = c^2$$<figcaption>Pythagoras Theorem.</figcaption></figure>"`;
